keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,202 @@
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Gt
+            values:
+            - 20
+          - key: foo
+            operator: Lt
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ ### Risks and Mitigations
+
+ Along with this feature, the biggest risk should be performance. This may also be the reason 
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. In order to understand the impact of these changes, we need to understand the their performance implications. We will perform performance/benchmark tests on this change and backward compatibility test.
+
+ Compared to the time that inter-pod affinity was introduced, performance of the scheduler has been greatly improved(https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly Optimized(https://github.com/kubernetes/kubernetes/pull/67788).
+
+ ## Design Details
+ ### Content
+  We will abstract the `Node` and `Pod` label selection feature, and finally the `NodeSelector` and `PodSelector` will be based on the `LabelSelectorRequirement` implementation:
+
+ The definition of `NodeSelector` will not change,it is defined as below:
+ ```go
+type NodeSelector struct {
+  NodeSelectorTerms []NodeSelectorTerm
+}
+```
+
+For the definition of `NodeSelectorTerm`, today the api looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []NodeSelectorRequirement
+  MatchFields []NodeSelectorRequirement
+}
+```
+
+We will use `LabelSelectorRequirement` to replace the original `NodeSelectorRequirement` as the base selector of `NodeSelector`. So it looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []LabelSelectorRequirement
+  MatchFields []LabelSelectorRequirement
+}
+```
+
+`PodSelector` is our new selector. A pod selector represents the union of the results of one or more label queries over a set of pods. The of `PodSelector` is defined as below:
+ ```go
+type PodSelector struct {
+  MatchLabels map[string]string
+  MatchExpressions []LabelSelectorRequirement
+}
+```
+
+`LabelSelectorRequirement` is an abstract selector implementation of `NodeSelector` and `PodSelector`. And the `LabelSelectorRequirement` is defined as below:
+ ```go
+type LabelSelectorRequirement struct {
+  Key string
+  Operator LabelSelectorOperator
+  Values []string
+}
+
+type LabelSelectorOperator string
+
+const (
+  LabelSelectorOpIn           LabelSelectorOperator = "In"
+  LabelSelectorOpNotIn        LabelSelectorOperator = "NotIn"
+  LabelSelectorOpExists       LabelSelectorOperator = "Exists"
+  LabelSelectorOpDoesNotExist LabelSelectorOperator = "DoesNotExist"
+  LabelSelectorOpGt           LabelSelectorOperator = "Gt"
```

Comment:
I think this name needs to indicate that the string-typed value will first be interpreted as a number; one could read this as meaning lexically greater--actually that's the natural meaning since label values are strings. Also, typically we request that things be spelled out in the API, although we should probably be consistent. :/ If this were starting from scratch I'd suggest the name "NumericallyGreater". You can still change the name of the constant to e.g. "LabelSelectorNumericallyGreater".

Reply:
DONE, Thank you very much for your help and corrections.